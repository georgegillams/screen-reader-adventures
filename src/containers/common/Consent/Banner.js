import React from 'react';
import PropTypes from 'prop-types';
import { cssModules } from 'gg-components/helpers/cssModules';
import STYLES from './banner.scss';

const getClassName = cssModules(STYLES);

const Banner = props => {
  const { children, className, ...rest } = props;

  return (
    <div className={getClassName('banner__outer', className)} {...rest}>
      <div className={getClassName('banner__inner')}>{children}</div>
    </div>
  );
};

Banner.propTypes = {
  children: PropTypes.object.isRequired,
  className: PropTypes.string,
};

Banner.defaultProps = {
  className: null,
};

export default Banner;
