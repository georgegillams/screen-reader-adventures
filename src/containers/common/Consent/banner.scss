@import '~utils/tokens';

.banner {
  &__outer {
    position: fixed;
    bottom: 0;
    left: 0;
    display: flex;
    z-index: 100000;
    width: 100vw;
    justify-items: center;
    justify-content: center;
    border-top: solid $alternating-background-color-1 0.15rem;
    background: $alternating-background-color-1;

    @media (prefers-color-scheme: dark) {
      border-top: solid white 0.15rem;
      background: $alternating-background-color-1-invert;
    }
  }

  &__inner {
    display: flex;
    width: 100%;
    max-width: 58rem;
    padding: 1rem;
    justify-content: space-around;
    align-items: center;
  }
}
