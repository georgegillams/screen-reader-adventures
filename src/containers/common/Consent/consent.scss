@import '~utils/tokens';

.consent {
  &__modal-overlay {
    position: fixed;
    top: 0;
    left: 0;
    display: flex;
    z-index: 1000;
    width: 100vw;
    height: 100vh;
    justify-content: center;
    align-items: center;
    background: rgba(8, 8, 8, 0.733);
  }

  &__modal-content {
    width: auto;
    max-width: calc(100vw - $spacing-lg);
    max-width: 30rem;
    padding: $spacing-lg;
    border: 0.15rem gray solid;
    border-radius: 0.4rem;
    background: $background-color;

    @include bpk-breakpoint-mobile {
      max-width: 80%;
    }

    @media (prefers-color-scheme: dark) {
      background: $background-color-invert;
    }
  }

  &__button {
    margin: 0.5rem 1rem 0 0;
  }

  &__section {
    &--banner {
      margin-right: $spacing-lg;
    }
    &--modal {
      margin-bottom: $spacing-lg;
    }
  }

  &__section-text {
    margin: 0 0 0.4rem 0;
  }
}
