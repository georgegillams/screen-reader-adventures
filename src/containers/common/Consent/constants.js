export const KEY = 'consent';
export const CONSENT = `app/${KEY}/CONSENT`;
export const DEFER_CONSENT = `app/${KEY}/DEFER_CONSENT`;
export const SET_CONSENT_REASON = `app/${KEY}/SET_CONSENT_REASON`;
export const RESET_CONSENT = `app/${KEY}/RESET_CONSENT`;

export const CONSENT_STATE_UNSET = 'CONSENT_STATE_UNSET';
export const CONSENT_STATE_ALLOWED = 'CONSENT_STATE_ALLOWED';
export const CONSENT_STATE_DEFERRED = 'CONSENT_STATE_DEFERRED';

export const CONSENT_STATE_ALLOWED_CLIENT_VALUE = 'CONSENT_ALLOWED';
export const CONSENT_STATE_DEFERRED_CLIENT_VALUE = 'CONSENT_DEFERRED';
