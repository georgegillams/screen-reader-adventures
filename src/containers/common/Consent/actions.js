import { createRoutine } from 'redux-saga-routines';

import { CONSENT, DEFER_CONSENT, SET_CONSENT_REASON, RESET_CONSENT } from './constants';

export const consent = createRoutine(CONSENT);
export const deferConsent = createRoutine(DEFER_CONSENT);
export const setConsentReason = createRoutine(SET_CONSENT_REASON);
export const resetConsent = createRoutine(RESET_CONSENT);
