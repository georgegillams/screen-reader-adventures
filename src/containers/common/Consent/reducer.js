import produce from 'immer';

import { consent, deferConsent, setConsentReason, resetConsent } from './actions';
import { CONSENT_STATE_UNSET, CONSENT_STATE_ALLOWED, CONSENT_STATE_DEFERRED } from './constants';

export const initialState = {
  cookieConsent: null,
  cookieConsentReason: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case resetConsent.TRIGGER:
        draft.cookieConsent = CONSENT_STATE_UNSET;
        break;

      case consent.TRIGGER:
        draft.cookieConsent = CONSENT_STATE_ALLOWED;
        break;

      case deferConsent.TRIGGER:
        draft.cookieConsent = CONSENT_STATE_DEFERRED;
        break;

      case setConsentReason.TRIGGER:
        draft.cookieConsentReason = payload;
        break;
    }
  });

export default reducer;
