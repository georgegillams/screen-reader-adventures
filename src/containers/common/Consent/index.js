import { memo } from 'react';

import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { consent, deferConsent, resetConsent } from './actions';
import { selectState } from './selectors';
import Consent from './Container';
import injectSaga from 'client-utils/common/redux/inject-saga';
import injectReducer from 'client-utils/common/redux/inject-reducer';

import { KEY } from './constants';
import saga from './saga';
import reducer from './reducer';

const mapStateToProps = createStructuredSelector({
  consentState: selectState(),
});

export function mapDispatchToProps(dispatch) {
  return {
    consent: payload => dispatch(consent(payload)),
    deferConsent: payload => dispatch(deferConsent(payload)),
    resetConsent: payload => dispatch(resetConsent(payload)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);
const withSaga = injectSaga({ key: KEY, saga });
const withReducer = injectReducer({ key: KEY, reducer });

export default compose(withSaga, withReducer, withConnect, memo)(Consent);
