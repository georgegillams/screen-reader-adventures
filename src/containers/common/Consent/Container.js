import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { DebugObject } from 'components/common/DebugObject';
import {
  CONSENT_STATE_UNSET,
  CONSENT_STATE_ALLOWED,
  CONSENT_STATE_ALLOWED_CLIENT_VALUE,
  CONSENT_STATE_DEFERRED_CLIENT_VALUE,
} from './constants';
import Button from 'components/common/Button';
import { Paragraph } from 'gg-components/Paragraph';
import Modal from 'react-modal';
import { cssModules } from 'gg-components/helpers/cssModules';
import STYLES from './consent.scss';
import { getPrivacyPreferences, setPrivacyPreferences } from 'client-utils/common/storageHelpers';
import TextLink from 'components/common/TextLink';
import Banner from './Banner';
import { Section } from 'gg-components/Section';

const getClassName = cssModules(STYLES);

const Consent = props => {
  const {
    consent,
    deferConsent,
    resetConsent,

    consentState,

    ...rest
  } = props;

  const test = process.env.NODE_ENV === 'test';

  useEffect(() => {
    if (!test) {
      Modal.setAppElement('#__next');
    }
    const privacyPreferences = getPrivacyPreferences();
    if (privacyPreferences === CONSENT_STATE_DEFERRED_CLIENT_VALUE) {
      deferConsent();
    } else if (privacyPreferences === CONSENT_STATE_ALLOWED_CLIENT_VALUE) {
      consent();
    } else {
      resetConsent();
    }
  }, []);

  let showBanner = false;
  let showModal = false;
  if (consentState.cookieConsent === CONSENT_STATE_UNSET && !consentState.cookieConsentReason) {
    showBanner = true;
  }
  if (consentState.cookieConsent !== CONSENT_STATE_ALLOWED && !!consentState.cookieConsentReason) {
    showModal = true;
  }

  const consentChildren = (
    <>
      <Section
        className={getClassName(
          'consent__section',
          !!showBanner && 'consent__section--banner',
          !!showModal && 'consent__section--modal'
        )}
        textClassName={getClassName('consent__section-text')}
        padding={false}
        name="Privacy and cookies">
        <Paragraph>
          To provide the best possible experience, we would like to use cookies and handle your data in accordance with
          our <TextLink href="/privacy-policy">Privacy policy</TextLink>.
          <br />
          {consentState.cookieConsentReason && (
            <>
              <br />
              You must accept usage of cookies and data to {consentState.cookieConsentReason}.
            </>
          )}
        </Paragraph>
      </Section>
      <div>
        <Button
          className={getClassName('consent__button')}
          onClick={() => {
            setPrivacyPreferences(CONSENT_STATE_ALLOWED_CLIENT_VALUE);
            consent();
          }}>
          Accept
        </Button>
        <Button
          className={getClassName('consent__button')}
          onClick={() => {
            deferConsent();
            setPrivacyPreferences(CONSENT_STATE_DEFERRED_CLIENT_VALUE);
          }}
          href={consentState.cookieConsentReason ? '/' : null}>
          {consentState.cookieConsentReason ? 'Dismiss and go to home page' : 'Dismiss'}
        </Button>
      </div>
    </>
  );

  let cookieConsentComponent = null;
  if (showBanner) {
    cookieConsentComponent = <Banner>{consentChildren}</Banner>;
  }
  if (showModal) {
    // Show modal
    cookieConsentComponent = (
      <Modal
        isOpen
        ariaHideApp
        overlayClassName={getClassName('consent__modal-overlay')}
        className={getClassName('consent__modal-content')}>
        {consentChildren}
      </Modal>
    );
  }

  return (
    <div {...rest}>
      <DebugObject
        debugTitle="Consent"
        debugObject={{
          consent,
          deferConsent,
          consentState,
        }}
      />
      {cookieConsentComponent && cookieConsentComponent}
    </div>
  );
};

Consent.propTypes = {
  test: PropTypes.bool,
  consent: PropTypes.func.isRequired,
  deferConsent: PropTypes.func.isRequired,
  resetConsent: PropTypes.func.isRequired,
  consentState: PropTypes.shape({
    cookieConsent: PropTypes.string,
    cookieConsentReason: PropTypes.string,
  }).isRequired,
};

Consent.defaultProps = {
  test: false,
};

export default Consent;
