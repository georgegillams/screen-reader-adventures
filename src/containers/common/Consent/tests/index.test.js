import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initialState } from '../reducer';

import configureStore from 'client-utils/common/redux/configure-store';

import ConsentIndex from '../index';
import Consent from '../Container';
import { CONSENT_STATE_ALLOWED, CONSENT_STATE_DEFERRED } from '../constants';

describe('<Consent />', () => {
  let store;
  const spy = jest.fn();

  beforeAll(() => {
    store = configureStore({});
  });

  it('should render correctly - index', () => {
    const { container } = render(
      <Provider store={store}>
        <ConsentIndex />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <Consent
          consent={spy}
          resetConsent={spy}
          deferConsent={spy}
          consentState={{
            ...initialState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with consent allowed', () => {
    const { container } = render(
      <Provider store={store}>
        <Consent
          consent={spy}
          resetConsent={spy}
          deferConsent={spy}
          consentState={{
            ...initialState,
            cookieConsent: CONSENT_STATE_ALLOWED,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with consent deferred', () => {
    const { container } = render(
      <Provider store={store}>
        <Consent
          consent={spy}
          resetConsent={spy}
          deferConsent={spy}
          consentState={{
            ...initialState,
            cookieConsent: CONSENT_STATE_DEFERRED,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with consent deferred and reason', () => {
    const { container } = render(
      <Provider store={store}>
        <Consent
          consent={spy}
          resetConsent={spy}
          deferConsent={spy}
          consentState={{
            ...initialState,
            cookieConsent: CONSENT_STATE_DEFERRED,
            cookieConsentReason: 'log in',
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
