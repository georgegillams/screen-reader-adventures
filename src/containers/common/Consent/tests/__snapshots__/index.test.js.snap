// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Consent /> should render correctly - index 1`] = `
<div>
  <div>
    <div
      class="banner__outer"
    >
      <div
        class="banner__inner"
      >
        <div
          class="section__outer consent__section consent__section--banner"
        >
          <h2
            class="section__heading section__heading--no-padding consent__section-text"
            id="privacy-and-cookies"
          >
            Privacy and cookies
          </h2>
          <span
            class="paragraph"
          >
            To provide the best possible experience, we would like to use cookies and handle your data in accordance with our 
            <a
              class="text-link text-link--text-link text-link--no-padding text-link--inline"
              href="/privacy-policy"
            >
              Privacy policy
            </a>
            .
            <br />
          </span>
        </div>
        <div>
          <button
            class="button__outer button__outer--regular button__outer--no-bouncy consent__button"
            type="button"
          >
            <span
              class="button__child"
            >
              Accept
            </span>
          </button>
          <button
            class="button__outer button__outer--regular button__outer--no-bouncy consent__button"
            type="button"
          >
            <span
              class="button__child"
            >
              Dismiss
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Consent /> should render correctly 1`] = `
<div>
  <div />
</div>
`;

exports[`<Consent /> should render correctly with consent allowed 1`] = `
<div>
  <div />
</div>
`;

exports[`<Consent /> should render correctly with consent deferred 1`] = `
<div>
  <div />
</div>
`;

exports[`<Consent /> should render correctly with consent deferred and reason 1`] = `
<div>
  <div />
</div>
`;
